Fixed #28295 - Removed trailing hyphens from ModelAdmin.prepopulated_fields

monotonee opened the pull request:
The JavaScript responsible for generating the output of ModelAdmin.prepopulated_fields would sometimes leave a trailing hyphen in the returned string. A single, extra operation was added to urlify.js to ensure that the resulting output will no longer end with a hyphen.

timgraham commented:
Could you try adding a test in `js_tests/admin`?

monotonee commented:
@timgraham I added a handful of tests for URLify.js. I apologize for their initial omission. I'd be glad to amend anything else you require. Given the increased number of contributed lines from the tests, would you like me to submit a CLA?

timgraham commented on js_tests/admin/URLify.test.js:
```diff
@@ -0,0 +1,29 @@
+});
+
+QUnit.test('merge adjacent whitespace', function(assert) {
+    // Adjacent whitespace should be merged prior to truncation.
```
should be -> are

timgraham commented on js_tests/admin/URLify.test.js:
```diff
@@ -0,0 +1,29 @@
+});
+
+QUnit.test('no post-truncation hyphen', function(assert) {
+    // Ensure that string truncation does not leave a superfluous trailing hyphen.
```
As suggested in [Coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style), state the expected behavior, e.g. "Trailing hyphen are removed." and avoid use of "Ensure that" since all tests ensure stuff.

timgraham commented on js_tests/admin/URLify.test.js:
```diff
@@ -0,0 +1,29 @@
+    assert.strictEqual(URLify('', 8, true), '');
+});
+
+QUnit.test('strip non-essential words', function(assert) {
```
no dash in "nonessential"

timgraham commented on js_tests/admin/URLify.test.js:
```diff
@@ -0,0 +1,29 @@
+    assert.strictEqual(URLify('the D is silent', 8, true), 'd-silent');
+});
+
+QUnit.test('strip non-URL-path characters', function(assert) {
```
non-URL path

timgraham commented on js_tests/admin/URLify.test.js:
```diff
@@ -0,0 +1,29 @@
+    assert.strictEqual(URLify('D   silent', 8, true), 'd-silent');
+});
+
+QUnit.test('no post-truncation hyphen', function(assert) {
```
no trailing hyphens

monotonee commented:
Thanks for taking the time to review the code. I effected the changes you required and I moved the passing unit tests to #8636. This pull request is now rebased from #8636 and is dependent upon it. Please let me know if I can be of further help.
